README
XmAddressBook - A Motif Address Book - Version 1.5.3

Copyright 1994,1995,1996 by Boyd Fletcher and Old Dominion University

Send problems/questions to: boyd@ccpo.odu.edu

NOTE: This used to be called XmRolodex.


DESCRIPTION:

XmAddressbook is a X11/Motif based Address Book program. It can convert the old
rolo program's rolodex.dat files into its own format. 

SUPPORTED SYSTEMS:

Company			Machine			OS
-------------------------------------------------------------------------
Sun Microsystems	SPARCstations		Solaris 1.x (SUNOS 4.1.x)
						using GCC
Sun Microsystems	SPARCstations		Solaris 2.x (SUNOS 5.x) 
IBM			RS/6000 		AIX 3.2.x and AIX 4.1
Silicon Graphics	Indigo 2		IRIX 5.x
DEC 			AXP 3000/600		OSF 2.0b
Solbourne		Series 5/600            OS/MP 4.1C using GCC
Hewlett Packard		HP 9000/735             HP-UX 9.X
Novell			Intel x86, Pentium	UnixWare 1.1 & UnixWare 2.0
BSDI			Intel x86, Pentium	BSDI 1.1
OpenVMS			Alpha & VAX		OpenVMS 6.1 and 6.2 (X11R5/Motif 1.2)

NOTE: I have tested XmAddressBook on the following SunOS, AIX, IRIX,
and OS/MP.

INSTALLATION: (except OpenVMS)

1) Edit Imakefile if you are using X11R6

2) If you have Imake, then type: xmkmf

3) type: make

4) type: make install

5) type: make install.man

OR,

1) type: cp Makefile.noimake Makefile

2) Edit the Makefile for your specific machine. Paying particular
attention to the following fields:

HELP_FILE=/usr/local/X11/lib/X11/xmaddressbook/xmaddressbook.hlp
BIN_DIR=/usr/local/X11/bin
MAN_DIR=/usr/local/X11/man/man1
APP_DEFAULTS_DIR=/usr/local/X11/lib/X11/app-defaults
APP_DEFAULTS=XmAddressbook
MAN_PAGE=xmaddressbook.1

3) type: make 

4) type: make install


OpenVMS:

Take a look at the DESCRIP.MMS (similar to Unix's imake program) and
MAKE.COM files. (thanks to Martin P.J. Zinser, m.zinser@gsi.de)

PORTING:

If you port it to another platform please send be the diffs and the
options used to compile it.

HOW TO GET IT:

Using anonymous ftp:

ftp ftp.ccpo.odu.edu

Connected to ftp.ccpo.odu.edu.
220 kuroshio FTP server (SunOS 4.1) ready.
Name (ftp.ccpo.odu.edu:boyd): ftp
331 Guest login ok, send ident as password.
Password:  (use your login name here)
230 Guest login ok, access restrictions apply.
ftp> cd pub/boyd
250 CWD command successful.
ftp> bin
200 Type set to I.
ftp> get xmaddressbook-Maj.Min.Rev.tar.Z

COPYING:

You are free to do anything you want with this program except sell it
or remove the above copyright.

DISCLAIMER:

BECAUSE THE PROGRAM IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW.  EXCEPT
WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER
PARTIES PROVIDE THE PROGRAM "AS IS" WITHOUT WARRANTY OF ANY KIND,
EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE.  THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE
PROGRAM IS WITH YOU.  SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME
THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR
DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL
DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE PROGRAM
(INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING RENDERED
INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A FAILURE OF
THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS), EVEN IF SUCH HOLDER
OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.








HISTORY
v1.5.3 05/11/96 - boyd

* Fixed glitch in Postscript printing routine that caused the wrong
  number of entries to be printed. (found by David_Barr@smtp.rc.trw.com)

v1.5.2 01/05/96 - Martin Zinser (m.zinser@gsi.de) and  boyd

* Added support for VMS (Zinser)
* changed -c to mean curse mode
* changed fonts and colors

------ 
v1.5.1 7/20/95 - boyd

* Fixed colors in name list, business card, and send mail routines

------ 
v1.5 7/7/95 - boyd

* Added a sending mail feature.
* Fixed bug in business card when there are no entries.
* Fixed text exceeds line in curses mode so that it wraps correctly
* Fixed auto raising when button pressed and window hidden.

------ 
v1.4 6/30/95 - boyd

* Added a curses interface for searching and viewing the addressbook.
  Makes those times when you are using a modem easier.

------ 
v1.3 6/25/95 - boyd

* Changed defaults fonts again (I thought the previous choices looked corney)
* created misc menu 
* added a show name list option to misc menu
* added search and goto to misc menu
* moved update button to misc menu
* changed the way goto works. you now get a popup. this is hopefully less 
  confusing
* changed print, search, and duplicate to popup dialogs 
* changed the CLEAR before ADD feature, you should still clear first, but 
  you can add without clearing.
* add a "business card" popup so that the whole card could be selected 
  for pasting (suggested  Forrest Aldrich)

------ 
v1.2.2 2/6/95 - boyd

* Added ranges to print routines
* Fixed file closing problem in print.c
* Fixed WM_DELETE_WINDOW problem in all popup windows
* Added Text output (print,file)
* Added Duplication feature (suggested by Michael Coxe)
* Fixed bug in print routine that caused new entries not be printed

------ 
v1.2.1 1/23/95 - boyd

* Changed name to XmAddressBook.
* Minor bug fixes.

------ 
v1.2 12/8/94 - boyd

* Added the ability for mulitple address book files.
* Improved error message handling.
* bug fixes

------ 
v1.1 11/29/94 - boyd

* Added Imakefile, changed date/time to locale format, removed
DEFAULT_ADDRESSBOOK_PATH and changed to $HOME (thanks to Richard Lloyd
<R.K.Lloyd@csc.liv.ac.uk>)
* Added Fax and Email entries. NOTE: You need to use the -convert1.1
option to convert your 1.0.X XmAddressBook files to 1.1 XmAddressBook Files.
* bug fixes

------ 
v1.0.4 11/28/94 - boyd 

* Changes to font handling

------ 
v1.0.2 10/28/94 - boyd

* Removed need to have PRINTER &/or LASER environment variables set.
* Improved search routines.

------ 
v1.0.1 10/15/94 - boyd

* Minor change to search routines
* Other minor bug fixes

------ 
v1.0 10/1/94 - boyd

* Initial Release!


mail.c
#include "main.h"

Widget Mail_Popup;
Widget MailTo_Text;
Widget MailCC_Text;
Widget MailBCC_Text;
Widget MailSubject_Text;
Widget MailAttachment_Text;
Widget MailMessage_Text;

void Open_Attachment_Dialog();




void Mail_Close(Widget   w,
		Widget   PopupWidget, 
		XmAnyCallbackStruct *call_value)
{
  G_MailVisible = False;
  XtUnrealizeWidget(PopupWidget);
}

void Mail_Confirm(Widget   w,
		  Widget   PopupWidget, 
		  XmAnyCallbackStruct *call_value)
{
  char *To,*CC,*BCC,*Subject,*Attachment,*Message;
  char *Attachment_String = NULL;
  FILE *PipeFile;
  FILE *AttachmentFile;
  struct stat AttachmentFileStat;
  char Buffer[MY_BUFSIZ];

  if (strlen(To = XmTextFieldGetString(MailTo_Text)) == 0)
    {
      ErrorMessage("Sending Mail Error:\n\nTO: Field empty. Mail not sent.\n");
      return;
    }

  CC = XmTextFieldGetString(MailCC_Text);
  BCC = XmTextFieldGetString(MailBCC_Text);
  Subject = XmTextFieldGetString(MailSubject_Text);
  Attachment = XmTextFieldGetString(MailAttachment_Text);
  Message = XmTextGetString(MailMessage_Text);

  if (strlen(Attachment) != 0)
    {
      if ((AttachmentFile = fopen(Attachment,"r")) == NULL)
	{
	  sprintf(Buffer,"File Error:\n\nCould not open Attachment file %s",Attachment);
	  ErrorMessage(Buffer); 
	  XtFree(To);
	  XtFree(CC);
	  XtFree(BCC);
	  XtFree(Subject);
	  XtFree(Attachment);
	  XtFree(Message);
	  return;
	}
      if (stat(Attachment,&AttachmentFileStat) != 0)
	{
	  sprintf(Buffer,"Sending Mail Error:\n\nCould not stat Attachment file %s",Attachment);
	  ErrorMessage(Buffer); 
	  XtFree(To);
	  XtFree(CC);
	  XtFree(BCC);
	  XtFree(Subject);
	  XtFree(Attachment);
	  XtFree(Message);
	  return;
	}
      Attachment_String = (char *)malloc(AttachmentFileStat.st_size+1);
      read(fileno(AttachmentFile),Attachment_String,AttachmentFileStat.st_size-1);
      Attachment_String[AttachmentFileStat.st_size+1] = '\0';
      Attachment_String[AttachmentFileStat.st_size] = '\0';
      fclose(AttachmentFile);
    }

  if ((PipeFile = popen(SENDMAIL_COMMAND,"w")) == NULL)
    {
      ErrorMessage("Sending Mail Error:\n\nCould not open mail process.\n");
      XtFree(To);
      XtFree(CC);
      XtFree(BCC);
      XtFree(Subject);
      XtFree(Attachment);
      XtFree(Message);
      if (strlen(Attachment) !=0)  free(Attachment_String);
      G_MailVisible = False;
      XtUnrealizeWidget(PopupWidget);
      return;
    }

  fprintf(PipeFile,"To: %s\n",To);
  if (strlen(CC) != 0 ) fprintf(PipeFile,"Cc: %s\n",CC);
  if (strlen(BCC) != 0) fprintf(PipeFile,"Bcc: %s\n",BCC);
  if (strlen(Subject) != 0) fprintf(PipeFile,"Subject: %s\n",Subject);
  fprintf(PipeFile,"X-Mailer: %s\n",PROGRAM_TITLE);
  if (strlen(Message) != 0) fprintf(PipeFile,"\n%s\n",Message);
  
  if (Attachment_String != NULL) 
    {
      fprintf(PipeFile,"\n--------- Attachment Begins ---------\n%s\n--------- Attachment Ends ---------\n",Attachment_String);
    }
  pclose(PipeFile);

  XtFree(To);
  XtFree(CC);
  XtFree(BCC);
  XtFree(Subject);
  XtFree(Attachment);
  XtFree(Message);
  free(Attachment_String);

  G_MailVisible = False;
  XtUnrealizeWidget(PopupWidget);
}

void Mail_Reset(Widget   w,
		     Widget   TextWidget, 
		     XmAnyCallbackStruct *call_value)
{

  XmTextFieldSetString(MailTo_Text,G_CurrentEntryInRoloList->EmailAddress);

  XmTextFieldSetString(MailCC_Text,"");
  XmTextFieldSetString(MailBCC_Text,"");
  XmTextFieldSetString(MailSubject_Text,"");
  XmTextFieldSetString(MailAttachment_Text,"");
  if (G_Signature_String != NULL)
    {
      XmTextSetString(MailMessage_Text,G_Signature_String);
      XmTextInsert(MailMessage_Text,0,"\n\n");
      XmTextSetCursorPosition(MailMessage_Text,0);
    }
  else {
    XmTextSetString(MailMessage_Text,"");
  }
}

void  Send_Mail_Popup()
{
  Widget Mail_Frame;
  Widget Mail_Form;
  Widget Mail_Form2;
  Widget MailTo_Label;
  Widget MailCC_Label;
  Widget MailBCC_Label;
  Widget MailSubject_Label;
  Widget MailAttachment_Label;
  Widget MailAttachment_Button;
  Widget Mail_Separator;
  Widget MailConfirm_Button;
  Widget MailReset_Button;
  Widget MailClose_Button;
  XmString LabelString;
  Atom WM_DELETE_WINDOW;

  if (G_RoloList == NULL)
    {
      return;
    }

  if (G_MailVisible == True)
    {
      XRaiseWindow(XtDisplay(Mail_Popup),XtWindow(Mail_Popup));
      XMapWindow(XtDisplay(Mail_Popup),XtWindow(Mail_Popup));
      return;
    }
  
  Mail_Popup = XtVaAppCreateShell("Mail",
				  "Mail_Popup",
				  applicationShellWidgetClass,
				  XtDisplay(W_TopLevel),
				  XmNtitle, "Mail",
				  XmNdeleteResponse, XmDO_NOTHING,
				  NULL,0); 
  
  Mail_Frame = XtVaCreateManagedWidget("Mail_Frame",
				       xmFrameWidgetClass, 
				       Mail_Popup,
				       NULL,0);
  
  Mail_Form = XtVaCreateManagedWidget("Mail_Form",
				      xmFormWidgetClass, 
				      Mail_Frame,
				      NULL,0);
  
  Mail_Form2 = XtVaCreateManagedWidget("Mail_Form2",
				       xmFormWidgetClass, 
				       Mail_Form,
				       XmNleftAttachment, XmATTACH_FORM,
				       XmNrightAttachment, XmATTACH_FORM,
				       XmNtopAttachment, XmATTACH_POSITION,
				       XmNtopPosition, 0,
				       XmNtopOffset,10,
				       XmNbottomAttachment, XmATTACH_POSITION,
				       XmNbottomPosition, 30,
				       NULL,0);
  
  LabelString = XmStringCreate("To: ", XmSTRING_OS_CHARSET),
  MailTo_Label = XtVaCreateManagedWidget("MailTo_Label",
					 xmLabelWidgetClass, 
					 Mail_Form2,
					 XmNleftOffset, 10,
					 XmNlabelString, LabelString,
					 XmNleftAttachment, XmATTACH_FORM,
					 XmNalignment, XmALIGNMENT_END,
					 XmNrightAttachment, XmATTACH_POSITION,
					 XmNrightPosition, 20,
					 XmNtopAttachment, XmATTACH_POSITION,
					 XmNtopPosition, 0,
					 XmNbottomAttachment, XmATTACH_POSITION,
					 XmNbottomPosition, 20,
					 NULL,0);
  XmStringFree(LabelString);
  MailTo_Text = XtVaCreateManagedWidget("MailTo_Text",
					xmTextFieldWidgetClass, 
					Mail_Form2,
					XmNrightOffset, 10,
					XmNleftAttachment, XmATTACH_POSITION,
					XmNleftPosition, 20,
					XmNrightAttachment, XmATTACH_FORM,
					XmNtopAttachment, XmATTACH_POSITION,
					XmNtopPosition, 0,
					XmNbottomAttachment, XmATTACH_POSITION,
					XmNbottomPosition, 20,
					NULL,0);
  
  XmTextFieldSetString(MailTo_Text,G_CurrentEntryInRoloList->EmailAddress);

  LabelString = XmStringCreate("CC: ", XmSTRING_OS_CHARSET),
  MailCC_Label = XtVaCreateManagedWidget("MailCC_Label",
					 xmLabelWidgetClass, 
					 Mail_Form2,
					 XmNleftOffset, 10,
					 XmNlabelString, LabelString,
					 XmNleftAttachment, XmATTACH_FORM,
					 XmNalignment, XmALIGNMENT_END,
					 XmNrightAttachment, XmATTACH_POSITION,
					 XmNrightPosition, 20,
					 XmNtopAttachment, XmATTACH_POSITION,
					 XmNtopPosition, 20,
					 XmNbottomAttachment, XmATTACH_POSITION,
					 XmNbottomPosition, 40,
					 NULL,0);
  XmStringFree(LabelString);
  MailCC_Text = XtVaCreateManagedWidget("MailCC_Text",
					xmTextFieldWidgetClass, 
					Mail_Form2,
					XmNrightOffset, 10,
					XmNleftAttachment, XmATTACH_POSITION,
					XmNleftPosition, 20,
					XmNrightAttachment, XmATTACH_FORM,
					XmNtopAttachment, XmATTACH_POSITION,
					XmNtopPosition, 20,
					XmNbottomAttachment, XmATTACH_POSITION,
					XmNbottomPosition, 40,
					NULL,0);
  
  
  LabelString = XmStringCreate("BCC: ", XmSTRING_OS_CHARSET),
  MailBCC_Label = XtVaCreateManagedWidget("MailBCC_Label",
					  xmLabelWidgetClass, 
					  Mail_Form2,
					  XmNleftOffset, 10,
					  XmNlabelString, LabelString,
					  XmNleftAttachment, XmATTACH_FORM,
					  XmNalignment, XmALIGNMENT_END,
					  XmNrightAttachment, XmATTACH_POSITION,
					  XmNrightPosition, 20,
					  XmNtopAttachment, XmATTACH_POSITION,
					  XmNtopPosition, 40,
					  XmNbottomAttachment, XmATTACH_POSITION,
					  XmNbottomPosition, 60,
					  NULL,0);
  XmStringFree(LabelString);
  MailBCC_Text = XtVaCreateManagedWidget("MailBCC_Text",
					 xmTextFieldWidgetClass, 
					 Mail_Form2,
					 XmNrightOffset, 10,
					 XmNleftAttachment, XmATTACH_POSITION,
					 XmNleftPosition, 20,
					 XmNrightAttachment, XmATTACH_FORM,
					 XmNtopAttachment, XmATTACH_POSITION,
					 XmNtopPosition, 40,
					 XmNbottomAttachment, XmATTACH_POSITION,
					 XmNbottomPosition, 60,
					 NULL,0);
  
  LabelString = XmStringCreate("Subject: ", XmSTRING_OS_CHARSET),
  MailSubject_Label = XtVaCreateManagedWidget("MailSubject_Label",
					      xmLabelWidgetClass, 
					      Mail_Form2,
					      XmNleftOffset, 10,
					      XmNlabelString, LabelString,
					      XmNleftAttachment, XmATTACH_FORM,
					      XmNalignment, XmALIGNMENT_END,
					      XmNrightAttachment, XmATTACH_POSITION,
					      XmNrightPosition, 20,
					      XmNtopAttachment, XmATTACH_POSITION,
					      XmNtopPosition, 60,
					      XmNbottomAttachment, XmATTACH_POSITION,
					      XmNbottomPosition, 80,
					      NULL,0);
  XmStringFree(LabelString);
  MailSubject_Text = XtVaCreateManagedWidget("MailSubject_Text",
					     xmTextFieldWidgetClass, 
					     Mail_Form2,
					     XmNrightOffset, 10,
					     XmNleftAttachment, XmATTACH_POSITION,
					     XmNleftPosition, 20,
					     XmNrightAttachment, XmATTACH_FORM,
					     XmNtopAttachment, XmATTACH_POSITION,
					     XmNtopPosition, 60,
					     XmNbottomAttachment, XmATTACH_POSITION,
					     XmNbottomPosition, 80,
					     NULL,0);
  
  LabelString = XmStringCreate("Attachment: ", XmSTRING_OS_CHARSET),
  MailAttachment_Label = XtVaCreateManagedWidget("MailAttachment_Label",
						 xmLabelWidgetClass, 
						 Mail_Form2,
						 XmNleftOffset, 10,
						 XmNlabelString, LabelString,
						 XmNleftAttachment, XmATTACH_FORM,
						 XmNalignment, XmALIGNMENT_END,
						 XmNrightAttachment, XmATTACH_POSITION,
						 XmNrightPosition, 20,
						 XmNtopAttachment, XmATTACH_POSITION,
						 XmNtopPosition, 80,
						 XmNbottomAttachment, XmATTACH_POSITION,
						 XmNbottomPosition, 100,
						 NULL,0);
  XmStringFree(LabelString);
  MailAttachment_Text = XtVaCreateManagedWidget("MailSubject_Text",
						xmTextFieldWidgetClass, 
						Mail_Form2,
						XmNleftAttachment, XmATTACH_POSITION,
						XmNleftPosition, 20,
						XmNtopAttachment, XmATTACH_POSITION,
						XmNtopPosition, 80,
						XmNbottomAttachment, XmATTACH_POSITION,
						XmNbottomPosition, 100,
						XmNcolumns, 50,
						NULL,0);

  LabelString = XmStringCreate("Browse...", XmSTRING_OS_CHARSET),
  MailAttachment_Button = XtVaCreateManagedWidget("MailAttachment_Button",
						  xmPushButtonWidgetClass, 
						  Mail_Form2,
						  XmNlabelString, LabelString,
						  XmNrightOffset, 10,
						  XmNleftAttachment, XmATTACH_WIDGET,
						  XmNleftWidget, MailAttachment_Text,
						  XmNrightAttachment, XmATTACH_FORM,
						  XmNtopAttachment, XmATTACH_POSITION,
						  XmNtopPosition, 80,
						  XmNbottomAttachment, XmATTACH_POSITION,
						  XmNbottomPosition, 100,
						  NULL,0);
  XmStringFree(LabelString);
  XtAddCallback(MailAttachment_Button, XmNactivateCallback, 
                (XtCallbackProc) Open_Attachment_Dialog, Mail_Popup);
  
  
  MailMessage_Text = XmCreateScrolledText(Mail_Form,
					  "MailMessage_Text",
					  NULL,0);
  
  XtVaSetValues(XtParent(MailMessage_Text),
		XmNleftAttachment, XmATTACH_FORM,
		XmNrightAttachment, XmATTACH_FORM,
		XmNrightOffset, 10,
		XmNleftOffset, 10,
		XmNtopAttachment, XmATTACH_POSITION,
		XmNtopPosition, 30,
		XmNbottomAttachment, XmATTACH_POSITION,
		XmNbottomPosition, 85,
		NULL,0);

  XtVaSetValues(MailMessage_Text,
		XmNcolumns, 80,
		XmNrows, 10,
		XmNwordWrap, True,
		XmNeditMode, XmMULTI_LINE_EDIT,
		NULL,0);

  XtManageChild(MailMessage_Text);
  
  
  if (G_Signature_String != NULL)
    {
      XmTextSetString(MailMessage_Text,G_Signature_String);
      XmTextInsert(MailMessage_Text,0,"\n\n");
      XmTextSetCursorPosition(MailMessage_Text,0);
    }
  
  Mail_Separator = XtVaCreateManagedWidget("Mail_Separator",
					   xmSeparatorWidgetClass,
					   Mail_Form,
					   XmNleftAttachment, XmATTACH_FORM,
					   XmNrightAttachment, XmATTACH_FORM,
					   XmNtopAttachment, XmATTACH_POSITION,
					   XmNtopPosition, 85,
					   XmNbottomAttachment, XmATTACH_POSITION,
					   XmNbottomPosition, 90,
					   NULL,0);
  
  LabelString = XmStringCreate("Send", XmSTRING_OS_CHARSET),
  MailConfirm_Button = XtVaCreateManagedWidget("MailConfirm_Button",
					       xmPushButtonWidgetClass,
					       Mail_Form,
					       XmNlabelString, LabelString,
					       XmNleftAttachment, XmATTACH_POSITION,
					       XmNleftPosition,25,
					       XmNtopAttachment, XmATTACH_POSITION,
					       XmNtopPosition, 90,
					       XmNbottomAttachment, XmATTACH_POSITION,
					       XmNbottomPosition, 95,
					       NULL,0);
  XmStringFree(LabelString);
  XtAddCallback(MailConfirm_Button, XmNactivateCallback, 
                (XtCallbackProc) Mail_Confirm, Mail_Popup);
  
  
  LabelString = XmStringCreate("Reset", XmSTRING_OS_CHARSET),
  MailReset_Button = XtVaCreateManagedWidget("MailReset_Button",
					     xmPushButtonWidgetClass,
					     Mail_Form,
					     XmNlabelString, LabelString,
					     XmNleftAttachment, XmATTACH_POSITION,
					     XmNleftPosition,50,
					     XmNtopAttachment, XmATTACH_POSITION,
					     XmNtopPosition, 90,
					     XmNbottomAttachment, XmATTACH_POSITION,
					     XmNbottomPosition, 95,
					     NULL,0);
  XmStringFree(LabelString);
  XtAddCallback(MailReset_Button, XmNactivateCallback, 
                (XtCallbackProc) Mail_Reset, NULL);
  
  
  LabelString = XmStringCreate("Cancel", XmSTRING_OS_CHARSET),
  MailClose_Button = XtVaCreateManagedWidget("MailClose_Button",
					     xmPushButtonWidgetClass,
					     Mail_Form,
					     XmNlabelString, LabelString,
					     XmNleftAttachment, XmATTACH_POSITION,
					     XmNleftPosition,75,
					     XmNtopAttachment, XmATTACH_POSITION,
					     XmNtopPosition, 90,
					     XmNbottomAttachment, XmATTACH_POSITION,
					     XmNbottomPosition, 95,
					     NULL,0);
  XmStringFree(LabelString);
  XtAddCallback(MailClose_Button, XmNactivateCallback, 
                (XtCallbackProc) Mail_Close, Mail_Popup );
  
  WM_DELETE_WINDOW = XmInternAtom(XtDisplay(Mail_Popup),"WM_DELETE_WINDOW",False);
  XmAddWMProtocolCallback(Mail_Popup,WM_DELETE_WINDOW,
                          (XtCallbackProc)Mail_Close,Mail_Popup);


/*   XtManageChild(Mail_Frame); */
/*   XtPopup(Mail_Popup,XtGrabExclusive); */
   XtRealizeWidget(Mail_Popup); 
  G_MailVisible = True;
}



void OK_Attachment(Widget   w,
		   Widget   PopupWidget, 
		   XmFileSelectionBoxCallbackStruct *cbs)
{
  char *Filename;

  if (!XmStringGetLtoR(cbs->value,XmSTRING_OS_CHARSET,&Filename))
    {
      XtUnrealizeWidget(w);
      return;      
    }
  if (Filename == NULL)
    {
      XtUnrealizeWidget(w); /* no filename given */
      XtFree(Filename);
      return;      
    }
  XmTextFieldSetString(MailAttachment_Text,Filename);
  XtFree(Filename);
  XtUnrealizeWidget(w);
  return;  
}

void CANCEL_Attachment(Widget   w,
			Widget   PopupWidget, 
			XmAnyCallbackStruct *cbs)
{
  XtUnrealizeWidget(w);
  return;
}

void Open_Attachment_Dialog()
{
  XmString DirString;
  Widget W_FileDialog;

  W_FileDialog = (Widget) XmCreateFileSelectionDialog(W_TopLevel,"Attachment",NULL,0);
  XtAddCallback(W_FileDialog, XmNcancelCallback, (XtCallbackProc) CANCEL_Attachment,NULL);
  XtAddCallback(W_FileDialog, XmNokCallback, (XtCallbackProc) OK_Attachment,NULL);
  XtUnmanageChild((Widget )XmFileSelectionBoxGetChild(W_FileDialog, XmDIALOG_HELP_BUTTON));
  XtVaSetValues(XtParent(W_FileDialog),
		XmNtitle,"Attachment File Selector",
		NULL,0);
  DirString = XmStringCreate((char *)getenv("HOME"), XmSTRING_OS_CHARSET),
  XtVaSetValues(W_FileDialog,
		XmNtextColumns, 60,
		XmNdirectory, DirString,
		NULL,0);
  XtManageChild(W_FileDialog);
}
